src/qpx_scan_algo.cpp

```diff
@@ -0,0 +1,940 @@
+
+  wait_unit_ready(dev, 6);
+  printf("Writing blocks: %ld - %ld (%ld MB)\n", lba_sta, lba_end,
+         (lba_end - lba_sta) >> 9);
+
+  dev->parms.write_speed_kb = speed * spd1X;
+  set_rw_speeds(dev);
+  get_rw_speeds(dev);
+  writer->setSimul(WT_simul);
+  if (dev->media.type & DISC_CD) {
+    printf("Running write transfer rate test on CD at speed %d...\n",
+           dev->parms.write_speed_kb / spd1X);
+    spd1X = 150;
+  } else if (dev->media.type & (DISC_DVD | DISC_BD)) {
+    printf("Running write transfer rate test on DVD/BD at speed %d...\n",
+           dev->parms.write_speed_kb / spd1X);
+  }
+
+  if (writer->open_session()) {
+    printf("Can't open session!\n");
+    goto write_cleanup;
+  }
+
+  if (writer->send_opc()) {
+    printf("OPC failed!\n");
+    goto write_cleanup;
+  }
+
+  if (writer->open_track(lba_end)) {
+    printf("Can't start new track!\n");
+    goto write_cleanup;
+  }
+
+  printf("Starting write...\n");
+  memset(dev->rd_buf, 0, bufsz_rd);
+
+  gettimeofday(&s, NULL);
+  gettimeofday(&blks, NULL);
+  for (lba = lba_sta; !stop_req && !err && lba < lba_end; lba += wsize) {
+    if (lba + wsize >= lba_end)
+      wsize = lba_end - lba;
+
+    get_wbuffer_capacity(dev, &ubuft, &ubuff);
+    //	if (ubuff >=0 && ubuff<(wsize<<11)) {
+    if ((ubuff >> 11) < ((uint32_t)wsize)) {
+      msleep(20);
+    }
+    if (wsize && writer->write_data(lba, wsize)) {
+      printf("\nWrite error at sector %d (wsize=%d)\n", lba, wsize);
+      stop_req = 1;
+    }
+    if (!(lba % bsize) || lba + wsize == lba_end || stat_req || stop_req ||
+        !wsize) {
+      ubufp = (uint32_t)(ubuft ? 100 * (1.0f - ubuff / (float)ubuft) : 0);
+      gettimeofday(&blke, NULL);
+      /*
+      btime=(blke.tv_sec - blks.tv_sec) + (blke.tv_usec -
+      blks.tv_usec)/1000000.0; spdKB = (bsize << 1) / btime; spdX  =
+      spdKB/(float)spd1X;
+      */
+      calc_cur_speed(((lba - 1) % bsize) + 1);
+      printf("lba: %7d    speed: %6.2f X  %6d kB/s, written: %4ldMB/%4ldMB, "
+             "Ubuf: %3d%%\r",
+             lba, spdX, spdKB, (lba - lba_sta) >> 9, (lba_end - lba_sta) >> 9,
+             ubufp);
+      gettimeofday(&blks, NULL);
+      stat_req = 0;
+#ifdef USE_FFLUSH
+      fflush(stdout);
+#endif
+    }
+  }
+  printf("\n");
+  gettimeofday(&e, NULL);
+  show_avg_speed(lba);
+  writer->close_track();
+
+  writer->fixate();
+  start_stop(dev, 0);
+  start_stop(dev, 1);
+  delete writer;
+  writer = NULL;
+  return 0;
+
+write_cleanup:
+  printf("Errors before writing! cleaning up...\n");
+  flush_cache(dev, true);
+  start_stop(dev, 0);
+  start_stop(dev, 1);
+  delete writer;
+  writer = NULL;
+  return 1;
+}
+
+#endif // #ifdef DISABLE_INTERNAL_WT
+
+//
+//  CD algo's
+//
+
+int qscanner::run_cd_errc() {
+  cd_errc err, err_tot, err_max;
+  int errc_data;
+  long lba = lba_sta;
+  long lbao;
+  if (!attached)
+    return -1;
+  if (!(dev->media.type & DISC_CD))
+    return 1;
+  lba = 0;
+  errc_data = plugin->errc_data();
```

Comment:
Codacy found an issue: [Variable 'errc_data' is assigned a value that is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)